### PR TITLE
Implement handler for storing PPS in localStorage upon SurveyDataTransferRequest

### PR DESCRIPTION
**Pull request by caroljli**
Stores PPS values of AnswerChoices in QuestionCatalog upon survey submission request to send as populated GPT to GAM. The format for sending the PPS values via tag follows this example:

```
{
      [googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]: 
        {values: ['12', '13', '14']},
      [googletag.enums.Taxonomy.IAB_CONTENT_2_2]:
        {values: ['22', '25', '57']},
}
```

### CODE REVIEW

File: src/proto/api_messages.ts
Comment (justinchou-google) [inline src/proto/api_messages.ts:1]: Seems like the formatter didn't run, can we restore these changes?

File: src/runtime/audience-action-flow.js [hunk 1]
```diff
@@ -340,6 +351,64 @@ export class AudienceActionFlow {
   }
 
   /**
+   * Populates localStorage with PPS configuration parameters based on
+   * SurveyDataTransferRequest.
+   * @param {SurveyDataTransferRequest} request
+   * @return {boolean}
+   * @private
+   **/
+
```
Comment (justinchou-google) [inline src/runtime/audience-action-flow.js:360]: nit: extra line

File: src/runtime/audience-action-flow.js [hunk 2]
```diff
@@ -340,6 +351,64 @@ export class AudienceActionFlow {
...
+      let existingIabTaxonomyMap;
+      if (!existingIabTaxonomy) {
+        existingIabTaxonomyMap = {
+          '[googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]': ppsConfigParams,
+        };
+      } else {
+        const existingIabMap = [
+          ...existingIabTaxonomy
+            .substring(existingIabTaxonomy.indexOf(':'))
+            .replace(/\D/g, ''),
```
Comment (justinchou-google) [inline src/runtime/audience-action-flow.js:386]: Can we make this into a function for clarity and unit testing?

File: src/runtime/audience-action-flow.js [hunk 3]
```diff
@@ -340,6 +351,64 @@ export class AudienceActionFlow {
...
+        )
+      );
+      const result = Promise.resolve(this.storage_.get(iabAudienceKey, true));
+      return await !!Promise.resolve(
+        this.storage_.get(iabAudienceKey, /* useLocalStorage= */ true)
+      );
+    } else {
+      // Answers with no PPS parameters should always evaluate to true with
+      // the feature flag enabled.
+      return true;
```
Comment (justinchou-google) [inline src/runtime/audience-action-flow.js:407]: nit: can we move this null check to the beginning to simplify this control sequence?

File: src/proto/api_messages.ts
Comment (justinchou-google) [reply src/proto/api_messages.ts:1]: seems like formatting was applied during migration: https://github.com/subscriptions-project/swg-js/commit/29658e96bac4141afb40d3f55494558843cc7c98#diff-ec8d982c9d32304c84c4f0f37b705b5b977b52215f9ad8095856126a6f217ee2R391
@ChrisAntaki , do we want to preserve these formatting changes where previously `api_messages.js` was just the raw output from the generator?

File: src/runtime/audience-action-flow.ts [hunk 4]
```diff
@@ -321,6 +330,69 @@ export class AudienceActionFlow {
   }
 
   /**
+   * Populates localStorage with PPS configuration parameters based on
+   * SurveyDataTransferRequest.
+   * @param {SurveyDataTransferRequest} request
+   * @return {boolean}
+   * @private
```
Comment (fikus) [inline src/runtime/audience-action-flow.ts:337]: Since this is TypeScript, types don't need to be specified in JSDoc. The `request` param can be removed from the doc (the type is in the param declaration). The return type and visibility (private) should be specified on the method itself.

File: src/runtime/audience-action-flow.ts [hunk 5]
```diff
@@ -321,6 +330,69 @@ export class AudienceActionFlow {
...
+      )
+    );
+    return await !!Promise.resolve(
+      this.storage_.get(iabAudienceKey, /* useLocalStorage= */ true)
+    );
+  }
+
+  /**
+   * Filters array to contain only an array of digits after the identifier key
+   * in localStorage.
```
Comment (fikus) [inline src/runtime/audience-action-flow.ts:387]: This looks inaccurate, the input is a string.

File: src/runtime/audience-action-flow.ts [hunk 6]
```diff
@@ -321,6 +330,69 @@ export class AudienceActionFlow {
...
+   * @param {Array} list
+   * @returns
+   */
```
Comment (fikus) [inline src/runtime/audience-action-flow.ts:390]: As above, the type information should be specified using TS syntax and can be removed from the JSDoc. Looks like this method is also meant to be private.

File: src/runtime/audience-action-flow.ts (hunk 4 above)
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:337]: sorry about that! the migration happened while i was still in the process of writing this PR which is why i had the annotations.

File: src/runtime/audience-action-flow.ts (hunk 6 above)
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:390]: sorry about that! migration happened mid-PR but i'll fix all the js->ts instances

File: src/runtime/audience-action-flow.ts [hunk 7]
```diff
@@ -296,7 +296,16 @@ export class AudienceActionFlow {
       this.storage_.storeEvent(StorageKeys.SURVEY_DATA_TRANSFER_FAILED);
     }
     const surveyDataTransferResponse = new SurveyDataTransferResponse();
-    surveyDataTransferResponse.setSuccess(dataTransferSuccess);
+    const isPpsEligible = request.getStorePpsInLocalStorage();
+
+    if (isPpsEligible) {
+      const configurePpsSuccess = await this.configureAnswerPpsData(request);
+      surveyDataTransferResponse.setSuccess(
+        configurePpsSuccess && dataTransferSuccess
```
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:304]: hmm ok that makes sense. i can put a client log in order to handle this which will be more useful but since it doesn't affect data transfer it makes sense that iterating on both tese variables is not wise.

File: src/runtime/audience-action-flow.ts [hunk 8]
```diff
@@ -321,6 +330,69 @@ export class AudienceActionFlow {
...
+      iabAudienceKey,
+      /* useLocalStorage= */ true
+    );
+    let existingIabTaxonomyMap;
+    if (!existingIabTaxonomy) {
+      existingIabTaxonomyMap = {
+        '[googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]': ppsConfigParams,
+      };
+    } else {
+      const existingIabMap = this.filterDigits_(existingIabTaxonomy);
```
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:366]: it's because the key is 'ppstaxonomies' but the value itself is stored in localstorage as one string `'[googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]': ppsConfigParams`, where only `ppsConfigParams` is the useful stringified array of numbers for us.

File: src/runtime/audience-action-flow.ts [hunk 9]
```diff
@@ -321,6 +330,69 @@ export class AudienceActionFlow {
...
+      };
+    }
+    await Promise.resolve(
+      this.storage_.set(
+        iabAudienceKey,
+        JSON.stringify(existingIabTaxonomyMap),
+        /* useLocalStorage= */ true
+      )
+    );
+    return await !!Promise.resolve(
```
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:380]: sounds good! will get of all return types.

File: src/runtime/audience-action-flow.ts [hunk 10]
```diff
@@ -321,6 +330,69 @@ export class AudienceActionFlow {
...
+    );
+    let existingIabTaxonomyMap;
+    if (!existingIabTaxonomy) {
+      existingIabTaxonomyMap = {
+        '[googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]': ppsConfigParams,
+      };
+    } else {
+      const existingIabMap = this.filterDigits_(existingIabTaxonomy);
+      existingIabTaxonomyMap = {
+        '[googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]': existingIabMap.concat(
```
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:368]: makes sense. added logic to remove.

File: src/runtime/audience-action-flow.ts [hunk 11]
```diff
@@ -321,6 +330,69 @@ export class AudienceActionFlow {
...
+    }
+
+    const existingIabTaxonomy = await this.storage_.get(
+      iabAudienceKey,
+      /* useLocalStorage= */ true
+    );
+    let existingIabTaxonomyMap;
+    if (!existingIabTaxonomy) {
+      existingIabTaxonomyMap = {
+        '[googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]': ppsConfigParams,
```
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:358]: yes, since the format takes on: 
```
googletag.setConfig({   pps: {     taxonomies: {      [googletag.enums.Taxonomy.IAB_AUDIENCE_1_1]:         {values: ['12', '13', '14']},      [googletag.enums.Taxonomy.IAB_CONTENT_2_2]:        {values: ['22', '25', '57']},    }  } });
--
```
i updated the key to accommodate the key for the values as defined the API.

File: src/runtime/audience-action-flow.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (oyj9109) [reply src/runtime/audience-action-flow.ts]: If we're storing something in a stringified object format, we should be able to parse it and get the value of object using the same key. In other words, for
```
itemToBeStored = {[PPS_KEY]: pps_values};
localStorage.setItem(STORAGE_KEY, JSON.stringify(itemToBeStored));
```
where `PPS_KEY` and `STORAGE_KEY` are string constants.

`JSON.parse(localStorage.getItem(STORAGE_KEY))` should be equal to `itemToBeStored` and thus `itemToBeStored[PPS_KEY]` is equal to `pps_values` which is an array of PPS string values.

File: src/runtime/audience-action-flow.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (oyj9109) [reply src/runtime/audience-action-flow.ts]: I think `concat` doesn't deal with de-duplication.

File: src/runtime/audience-action-flow.ts [hunk 12]
```diff
@@ -321,6 +327,68 @@ export class AudienceActionFlow {
...
+    if (ppsConfigParams.length === 0 || !ppsConfigParams) {
+      log(`[swg.js] No PPS values found in survey answer choices.`);
+    }
+
+    const existingIabTaxonomy = await this.storage_.get(
+      iabAudienceKey,
+      /* useLocalStorage= */ true
+    );
+    let ppsConfigParamsValues;
+    if (!existingIabTaxonomy) {
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:354]: Can we define a variable for the existing pps values (an array of string, which can be empty in case there's nothing) and append the current values to the array to construct a new array to override stored value without having to deal with if/else statements?

File: src/runtime/audience-action-flow.ts [hunk 13]
```diff
@@ -321,6 +327,68 @@ export class AudienceActionFlow {
...
+      ppsConfigParamsValues = {values: ppsConfigParams};
+    } else {
+      const existingIabMap = this.filterDigits(existingIabTaxonomy);
+      ppsConfigParamsValues = {
+        values: existingIabMap.concat(ppsConfigParams as string[]),
+      };
+      await Promise.resolve(
+        this.storage_.remove(iabAudienceKey, /* useLocalStorage= */ true)
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:362]: Do we need to remove the existing item? Can we just execute `this.storage_.set` to overwrite?

File: src/runtime/audience-action-flow.ts (hunk 12 above)
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:354]: makes sense. will do that

File: src/runtime/audience-action-flow.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts]: This makes sense if we were calling the localStorage.getItem() methods (verified in console for InsideNova) but the storage_ API returns a string and therefore returns a parse error when this is fetched.

File: src/runtime/audience-action-flow.ts [hunk 14]
```diff
@@ -321,6 +327,68 @@ export class AudienceActionFlow {
...
+  private async configureAnswerPpsData(
+    request: SurveyDataTransferRequest
+  ): Promise<void> {
+    const iabAudienceKey = StorageKeys.PPS_TAXONOMIES;
+    // PPS value field is optional and category may not be populated
+    // in accordance to IAB taxonomies.
+    const ppsConfigParams = request
+      .getSurveyQuestionsList()!
+      .flatMap((question) => question.getSurveyAnswersList())
+      .map((answer) => answer?.getPpsValue())
```
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:342]: the linter wouldn't compile without the `?` operator

File: src/runtime/audience-action-flow.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (caroljli) [reply src/runtime/audience-action-flow.ts]: ah I see. didn't know wrapping in square brackets would omit the default variable value. thanks!

File: src/runtime/audience-action-flow.ts [hunk 15]
```diff
@@ -321,6 +327,47 @@ export class AudienceActionFlow {
...
+      .filter((ppsValue) => ppsValue !== null);
+
+    const existingIabTaxonomy = await this.storage_.get(
+      iabAudienceKey,
+      /* useLocalStorage= */ true
+    );
+    const existingIabTaxonomyValues = existingIabTaxonomy
+      ? JSON.parse(existingIabTaxonomy)[Constants.PPS_AUDIENCE_TAXONOMY_KEY]
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:350]: nit: JSON.parse should be wrapped by try-catch for error handling.

Parsing an item stored in localStorage can throw a runtime error if the item is not in the expected format. If it fails to parse, we can just fallback the existing values to an empty array and move forward with the current values.

File: src/runtime/audience-action-flow.ts [hunk 16]
```diff
@@ -321,6 +327,47 @@ export class AudienceActionFlow {
...
+      .getSurveyQuestionsList()!
+      .flatMap((question) => question.getSurveyAnswersList())
+      .map((answer) => answer?.getPpsValue())
+      .filter((ppsValue) => ppsValue !== null);
+
+    const existingIabTaxonomy = await this.storage_.get(
+      iabAudienceKey,
+      /* useLocalStorage= */ true
+    );
+    const existingIabTaxonomyValues = existingIabTaxonomy
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:349]: nit: use optional chaining to avoid a runtime error when the expected key doesn't exist.

`JSON.parse(existingIabTaxonomy)?.[Constants.PPS_AUDIENCE_TAXONOMY_KEY]?.values` || []

(You don't need to nullness of `existingIabTaxonomy` as `JSON.parse(null)` evalues to `null` and it will be nicely handled by optional chaining.)

File: src/runtime/audience-action-flow.ts [hunk 17]
```diff
@@ -321,6 +327,47 @@ export class AudienceActionFlow {
...
+      ? JSON.parse(existingIabTaxonomy)[Constants.PPS_AUDIENCE_TAXONOMY_KEY]
+          .values
+      : [];
+    const iabTaxonomyValues = Array.from(
+      new Set(ppsConfigParams.concat(existingIabTaxonomyValues))
+    );
+    const existingIabTaxonomyMap = {
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:356]: nit: it's not existing, but something new to be store. What about `iabTaxonomy` or `ppsTaxonomy`?

File: src/runtime/audience-action-flow.ts [hunk 18]
```diff
@@ -321,6 +327,47 @@ export class AudienceActionFlow {
   }
 
   /**
+   * Populates localStorage with PPS configuration parameters based on
+   * SurveyDataTransferRequest.
+   **/
+  private async configureAnswerPpsData(
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:333]: nit: what about `storePpsValuesFromSurveyAnswers`?

File: src/runtime/audience-action-flow-test.js [hunk 19]
```diff
@@ -1010,6 +1046,105 @@ describes.realWin('AudienceActionFlow', (env) => {
...
+    const audienceActionFlow = new AudienceActionFlow(runtime, {
+      action: 'TYPE_REWARDED_SURVEY',
+      configurationId: 'configId',
+      onCancel: onCancelSpy,
+      autoPromptType: AutoPromptType.CONTRIBUTION,
+    });
+    activitiesMock.expects('openIframe').resolves(port);
+
+    const existingIabTaxonomyMap = {
+      [Constants.PPS_AUDIENCE_TAXONOMY_KEY]: {values: ['3', '4']},
```
Comment (oyj9109) [inline src/runtime/audience-action-flow-test.js:1092]: Can you add a value that's not in new to existing so that the test covers adding (de-duped) values to existing one?

**Review by oyj9109 — COMMENTED** [review]
Left a few more comments for nit. Overall looks very good and I really like how neat it ended up.

File: src/runtime/audience-action-flow.ts (hunk 15 above)
Comment (caroljli) [reply src/runtime/audience-action-flow.ts:350]: ahh that makes sense. would be super helpful, added :)

File: src/runtime/audience-action-flow.ts (hunk 16 above)
Comment (caroljli) [reply src/runtime/audience-action-flow.ts]: makes sense. forgot about optional chaining since the file migrated to ts haha

File: src/runtime/audience-action-flow.ts [hunk 20]
```diff
@@ -321,6 +327,50 @@ export class AudienceActionFlow {
...
+        JSON.parse(existingIabTaxonomy)?.[Constants.PPS_AUDIENCE_TAXONOMY_KEY]
+          ?.values || [];
+      const iabTaxonomyValues = Array.from(
+        new Set(ppsConfigParams.concat(existingIabTaxonomyValues))
+      );
+      const iabTaxonomy = {
+        [Constants.PPS_AUDIENCE_TAXONOMY_KEY]: {values: iabTaxonomyValues},
+      };
+
+      await Promise.resolve(
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:360]: Can we take it out of try/catch so that the current pps values are still stored even if reading/parsing existing signals fails?

Comment (caroljli) [reply src/runtime/audience-action-flow.ts:365]: yeah, ended up making it fall back to a default value

**Comment by caroljli** [comment]
fixing unit tests

File: src/runtime/audience-action-flow-test.js (hunk 19 above)
Comment (oyj9109) [reply src/runtime/audience-action-flow-test.js]: To be more specific with an example, I meant something like
- existing : 3, 4, 5
- new: 1, 2, 3
- expect: 1, 2, 3, 4, 5

File: src/runtime/audience-action-flow-test.js [hunk 21]
```diff
@@ -148,6 +174,16 @@ describes.realWin('AudienceActionFlow', (env) => {
     sandbox.stub(runtime, 'win').returns(winWithNoGtag);
   }
 
+  async function addToLocalStorage(key, value) {
+    await runtime.storage().set(key, value, true);
```
Comment (oyj9109) [inline src/runtime/audience-action-flow-test.js:178]: Can we set up an expectation of the return value instead of actually interacting a local storage instance which is suppoed to be mocked?

```
      storageMock
        .expects('get')
        .withExactArgs(...)
        .resolves(...)
        .once();
```

File: src/runtime/audience-action-flow.ts [hunk 22]
```diff
@@ -321,6 +327,55 @@ export class AudienceActionFlow {
...
+
+    const existingIabTaxonomy = await this.storage_.get(
+      iabAudienceKey,
+      /* useLocalStorage= */ true
+    );
+    let existingIabTaxonomyValues: string[] = [];
+    try {
+      const parsedExistingIabTaxonomyValues =
+        JSON.parse(existingIabTaxonomy)?.[Constants.PPS_AUDIENCE_TAXONOMY_KEY]
+          ?.values || [];
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.ts:353]: nit: `|| []` part is unnecessary as type is checked below.

File: package-lock.json
Comment (oyj9109) [inline package-lock.json:1]: Can you revert this change if unnecessary?

**Review by oyj9109 — APPROVED** [review]
LGTM, Great work!
